lunamark/reader/markdown.lua

```diff
@@ -1312,27 +1290,14 @@ function M.new(writer, options)
                                                      writer.string(infostring))
                          end
 
-
-  larsers.FencedDivBegin = larsers.fenced_div_begin
-                         * Cmt(P(true), function(_, i)
-                             -- keep track how deep nested we are
-                             div_level = div_level + 1
-                             return i
-                           end)
-
-  larsers.FencedDivEnd   = larsers.fenced_div_end
-                         * Cmt(P(true), function(_, i)
-                             local result
-                             -- check whether we are at top level and decrement
-                             -- div_level otherwise
-                             if div_level <= 0 then
-                               result = nil
-                             else
-                               div_level = div_level - 1
-                               result = i
-                             end
-                             return result
-                           end)
+    larsers.FencedDiv = P{ (larsers.fenced_div_begin)
+                           * C(((parsers.any - (larsers.fenced_div_begin + larsers.fenced_div_end)) + V(1))^0)
```

Comment:
How does this grammar handle edge cases such as the following:

````` md
::: {.some-classname}
This should not parse as a div.
``` md
:::
```
`````

Reply:
It breaks here, yes. The edge-case is dubious (how likely people are going to do this?), but granted, it's not what we would expect. As said, I have lacked time to check edge cases. And I do lack time to advance as fast you wanted, so I pushed what I had started.

It spits out:
~~~
<div class="some-classname">This should not parse as a div.

<p>``` md</p></div>

<p>```</p>
~~~

This is bad. **But** your own version now on master, unless mistaken, spits out:
~~~
<div class="some-classname">

This should not parse as a div.

<pre><code class="language-md">:::
</code></pre>
~~~

Which is not better (and even lacks the closing div?). Or did I miss something?

Reply:
Wouldn't something like this solve the fenced code block case:
```
    larsers.fencedcodeblock = parsers.TildeFencedCodeBlock + parsers.BacktickFencedCodeBlock
    larsers.FencedDiv = P{ (larsers.fenced_div_begin)
                           * C((((parsers.any - larsers.fencedcodeblock - (larsers.fenced_div_begin + larsers.fenced_div_end))) + V(1) + larsers.fencedcodeblock)^0)
                           * larsers.fenced_div_end } / function (attr, div)
                                                          -- The \n here are to force paragraphs in the ..div
                                                          -- Not sure Pandoc would generate Para nodes, rather than
                                                          -- Plain, though!
                                                          return parse_blocks("\n"..div.."\n"), attr
                                                        end / writer.div 
                                                        ...
```
(without considering there to cancel `larsers.fencedcodeblock` if the corresponding option is not enabled)
Or am I totally off ! (Which is possible too... LPeg is hard to grasp sometimes!)

Reply:
I think you are on the right track.

Reply:
> Or did I miss something?

Can you please add https://github.com/jgm/lunamark/pull/52#discussion_r1030976336 and https://github.com/jgm/lunamark/issues/53 to the unit tests, so that we can see how the code is failing at the moment and fix it? I don't have a local environment,where I could test lunamark.